### ISSUE TITLE
harbor-core timeouts chartmuseum API request when there's is a considerable amount of charts. 

### ISSUE THREAD

**Issue by fivezerosix** [goharbor/harbor#10016]
**Expected behavior and actual behavior:**
When there a considerable amount of charts and chart versions (around +10000) the components start behaving unexpectedly. On the logs, we have been able to capture events on Core and Chartmuseum (see below) which some users report errors while using helm commands against the specific repository. 

Also on the Platform UI, when listing the project (either the project with the +100000 charts is in the current page or you search for it) you will get a spinning wheel that does not finish loading the results. A workaround to this issue is to search for the specific project (will cause a spinning wheel issue and await for the error notification on the top of the UI) then delete a character on the search bar and then add the character again, this will put the result on the page. 

Once you are on the charts page, you will see that the pagination is incorrect as it shows something like the image below ( supposedly only showing the first 15 charts)

![image](https://user-images.githubusercontent.com/33496047/69679870-dd17e880-106e-11ea-90ee-758620af4f13.png)

But it does list every chart on the project. and if you move to the next page, you will get the same results.  

**Versions:**
Please specify the versions of following systems.

- harbor version:  1.8.0

**Additional context:**

- **Log files:** 

----------------------

CHARTMUSEUM LOGS 

-----------------------

2019/11/05 17:45:16 [Recovery] 2019/11/05 - 17:45:16 panic recovered:
GET /yourprojectname/index.yaml HTTP/1.1
Host: harbor-harbor-chartmuseum
Accept-Encoding: gzip
Authorization: Basic Y2hhcnRfY29udHJvbGxlcjo5eDhhVzZLd1pPYzZINEY0
User-Agent: Go-http-client/1.1
write tcp 10.42.5.25:9999->10.42.4.36:41616: write: broken pipe
/usr/local/go/src/runtime/panic.go:513 (0x42cc88)
/go/src/github.com/helm/chartmuseum/vendor/github.com/gin-gonic/gin/context.go:714 (0xd8789f)
/go/src/github.com/helm/chartmuseum/vendor/github.com/gin-gonic/gin/context.go:792 (0xd87fd0)
/go/src/github.com/helm/chartmuseum/pkg/chartmuseum/server/multitenant/handlers.go:93 (0xe73d75)
/go/src/github.com/helm/chartmuseum/pkg/chartmuseum/server/multitenant/routes.go:34 (0xe79b43)
/go/src/github.com/helm/chartmuseum/pkg/chartmuseum/router/router.go:194 (0xe221d4)
/go/src/github.com/helm/chartmuseum/pkg/chartmuseum/router/router.go:140 (0xe22fb3)
/go/src/github.com/helm/chartmuseum/vendor/github.com/gin-gonic/gin/context.go:108 (0xd84582)
/go/src/github.com/helm/chartmuseum/vendor/github.com/zsais/go-gin-prometheus/middleware.go:364 (0xe1f8fe)
/go/src/github.com/helm/chartmuseum/vendor/github.com/gin-gonic/gin/context.go:108 (0xd84582)
/go/src/github.com/helm/chartmuseum/vendor/github.com/gin-contrib/size/size.go:88 (0xdce1f4)
/go/src/github.com/helm/chartmuseum/vendor/github.com/gin-gonic/gin/context.go:108 (0xd84582)
/go/src/github.com/helm/chartmuseum/pkg/chartmuseum/router/middleware.go:48 (0xe22962)
/go/src/github.com/helm/chartmuseum/vendor/github.com/gin-gonic/gin/context.go:108 (0xd84582)
/go/src/github.com/helm/chartmuseum/vendor/github.com/gin-gonic/gin/recovery.go:48 (0xd95959)
/go/src/github.com/helm/chartmuseum/vendor/github.com/gin-gonic/gin/context.go:108 (0xd84582)
/go/src/github.com/helm/chartmuseum/vendor/github.com/gin-gonic/gin/gin.go:397 (0xd8cdd7)
/go/src/github.com/helm/chartmuseum/vendor/github.com/gin-gonic/gin/gin.go:390 (0xd8cb0b)
/go/src/github.com/helm/chartmuseum/vendor/github.com/gin-gonic/gin/gin.go:326 (0xd8c551)
/usr/local/go/src/net/http/server.go:2741 (0x6a4a0a)
/usr/local/go/src/net/http/server.go:1847 (0x6a0c05)

----------------------

CORE LOGS 

-----------------------

2019/11/05 18:12:12 [D] [server.go:2619] | 10.3.58.228| 500 |10.120998509s| match| GET /api/chartrepo/yourprojectname/charts r:/api/chartrepo/:repo/charts
2019-11-05T18:12:12Z [ERROR] [/chartserver/client.go:124]: GET '/api/yourprojectname/charts' failed with error: Get http://harbor-harbor-chartmuseum/api/yourprojectname/charts: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2019-11-05T18:12:12Z [ERROR] [/common/api/base.go:196]: fail to list charts: Get http://harbor-harbor-chartmuseum/api/yourprojectname/charts: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2019-11-05T18:12:12Z [ERROR] [/common/api/base.go:229]:
2019-11-05T18:12:12Z [ERROR] [/common/api/base.go:69]: GET /api/chartrepo/yourprojectname/charts failed with error: {"code":500,"message":"Internal Server Error"}
2019/11/05 18:12:12 [D] [server.go:2619] | 10.3.58.228| 500 |10.137439993s| match| GET /api/chartrepo/yourprojectname/charts r:/api/chartrepo/:repo/charts

Please let me know if there's anything else I can provide. 
Thanks.

**Comment by heww** [comment]
Which storage do you use, local file, s3 or something else? And what is the name of project to access?

**Comment by fivezerosix** [comment]
Hello! 

I'm using NFS and why is the project name relevant in this case? 

Thanks

**Comment by heww** [comment]
There is a limitation for chartmuseum #9759, hmmm,  NFS do not touch the limitation.

We will do a test for this issue asap.

**Comment by steven-zou** [comment]
@heww @ninjadq 

Shall we consider to fix it?

**Comment by xiongkun01** [comment]
@heww @ninjadq

Shall we consider to fix it?

**Comment by abcsnowwolf** [comment]
@xiongkun01 hi, have your fix it?

**Comment by RafalMaleska** [comment]
is there a fix for this issue? 
or a workaround?

**Comment by jcambronero** [comment]
We ran into this same issue while upgrading Harbor from v1.8 to v1.9. 
These are the habor-core logs. The pod was crashing and restarting over and over.

```
2020-07-04T04:29:35Z [INFO] [/core/config/config.go:74]: init secret store
2020-07-04T04:29:35Z [INFO] [/core/config/config.go:77]: init project manager based on deploy mode
2020-07-04T04:29:35Z [INFO] [/core/config/config.go:146]: initializing the project manager based on local database...
2020-07-04T04:29:35Z [INFO] [/core/main.go:178]: configurations initialization completed
2020-07-04T04:29:35Z [INFO] [/common/dao/base.go:84]: Registering database: type-PostgreSQL host-<postgres_server_FQDN> port-5432 database-core sslmode-"disable"
2020-07-04T04:29:35Z [INFO] [/common/dao/base.go:89]: Register database completed
2020-07-04T04:29:35Z [INFO] [/common/dao/pgsql.go:119]: Upgrading schema for pgsql ...
2020-07-04T04:29:35Z [INFO] [/common/dao/pgsql.go:122]: No change in schema, skip.
2020-07-04T04:29:35Z [INFO] [/common/config/encrypt/encrypt.go:60]: the path of key used by key provider: /etc/core/key
2020-07-04T04:29:35Z [INFO] [/core/main.go:80]: User id: 1 already has its encrypted password.
2020-07-04T04:29:35Z [INFO] [/chartserver/cache.go:184]: Enable redis cache for chart caching
2020-07-04T04:29:35Z [INFO] [/chartserver/reverse_proxy.go:58]: New chart server traffic proxy with middlewares
2020-07-04T04:29:35Z [INFO] [/core/api/chart_repository.go:607]: API controller for chart repository server is successfully initialized
2020-07-04T04:29:35Z [INFO] [/common/dao/base.go:64]: initialized clair database
2020-07-04T04:29:35Z [INFO] [/core/main.go:226]: initializing notification...
2020-07-04T04:29:35Z [INFO] [/pkg/notification/notification.go:50]: notification initialization completed
2020-07-04T04:29:35Z [INFO] [/core/main.go:248]: Because SYNC_REGISTRY set false , no need to sync registry
2020-07-04T04:29:35Z [INFO] [/core/main.go:251]: Init proxy
2020-07-04T04:29:35Z [ERROR] [/core/main.go:259]: Failed to parse SYNC_QUOTA: strconv.ParseBool: parsing "": invalid syntax
2020-07-04T04:29:35Z [INFO] [/core/main.go:107]: Start to sync quota data .....
2020-07-04T04:29:35Z [INFO] [/core/api/quota/migrator.go:90]: [Quota-Sync]:: start to ping server ... [chart]
2020-07-04T04:29:35Z [INFO] [/core/api/quota/migrator.go:95]: [Quota-Sync]:: success to ping server ... [chart]
2020-07-04T04:29:35Z [INFO] [/core/api/quota/migrator.go:96]: [Quota-Sync]:: start to dump data from server ... [chart]
2020-07-04T04:29:35Z [INFO] [/chartserver/cache.go:184]: Enable redis cache for chart caching
2020-07-04T04:29:35Z [INFO] [/core/api/quota/chart/chart.go:218]: API controller for chart repository server is successfully initialized
2020-07-04T04:29:45Z [INFO] [/core/api/quota/migrator.go:99]: [Quota-Sync]:: fail to dump data from server ... [chart], quit sync ...
2020-07-04T04:29:45Z [ERROR] [/core/main.go:109]: Fail to sync quota data, quota sync error on getting info of project: get content failed: send request GET /api/<project_name>/charts failed: Get http://harbor-harbor-chartmuseum/api/<project_name>/charts: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-07-04T04:29:45Z [FATAL] [/core/main.go:264]: quota migration error, quota sync error on getting info of project: get content failed: send request GET /api/<project_name>/charts failed: Get http://harbor-harbor-chartmuseum/api/<project_name>/charts: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```

We had to rollback the upgrade and now we're looking forward for a solution or a workaround to this issue.

**Comment by steven-zou** [comment]
As @xiongkun01 mentioned, this perf issue is inherited from the upstream project and then we may need more time to do investigations and check if we can have a fix at the harbor side.

**Comment by steven-zou** [comment]
As @xiongkun01 mentioned, this perf issue is inherited from the upstream project and then we may need more time to do investigations and check if we can have a fix at the harbor side.

**Comment by steven-zou** [comment]
It seems Chartmusuem has improved the `GetObjectSliceDiff ` logic @xiongkun01 mentioned above (O(n^2) to O(n)). So, possible solutions for this issue:

- upgrade Harbor to latest version
- workaround: use the latest chartmuseum image to instead the old one (the new image should be compatible with the older Harbor version as the chartmuseum related APIs are kept stable)

**Comment by jcambronero** [comment]
Hi @steven-zou, we tried the workaround today: 
- We attempted the upgrade from v1.8 to v1.9 again.
- Manually changed the chartmuseum image to the latest version (v2.0.1) and we left the rest of the components as-is.
Unfortunately, we're having the exact same issue.

**Comment by steven-zou** [comment]
@jcambronero 

Thanks a lot for the verification. Unfortunately, we found the improved code is not contained in their latest release. It's only at the master branch, check it here:

master: https://github.com/chartmuseum/storage/blob/7feb4823f486829bd521ecf6ca31b9863601975e/storage.go#L57
release(v0.8.0): https://github.com/chartmuseum/storage/blob/e8f06997d76ad955c62c49a211a43f3ac252e8c9/storage.go#L57

A potential possible way, as a workaround, we might try to build the latest ChartMuseum image that contains the improved code mention above (using master code).

**Comment by steven-zou** [comment]
I‘m afraid that even we have a newly build ChartMuseum that contains the improved code and may have a better perf, the chart getting method may still consume a less long time when there are large scale charts at the backend storage. The time may still be longer than the current timeout settings (10s) of HTTP client used at the Harbor side.

A safe and reasonable fix solution to this issue is also adjusting the HTTP client timeout settings. That may need to be introduced in a Harbor patch release. The next one is 2.0.2 which may be out at the end of July.

**Comment by steven-zou** [comment]
More references put here:

- issues at Chartmuseum side: https://github.com/helm/chartmuseum/issues/332
- PR for fixing such perf issue at Chartmuseum side: https://github.com/helm/chartmuseum/pull/339

**Comment by renmaosheng** [comment]
@heww we should also cover our helm as OCI for this scalability scenario. I guess the OCI registry will be better than Chart Museum?

**Comment by heww** [comment]
> @heww we should also cover our helm as OCI for this scalability scenario. I guess the OCI registry will be better than Chart Museum?

Dealing with helm chart as OCI artifact is no different dealing with a docker image, they are both follow the spec of [the oci distribution](https://github.com/opencontainers/distribution-spec/blob/master/spec.md), so we don't need to cover helm as oci for the scalability scenario.

Yes, the OCI registry will be better than ChartMuseum because the data need to render in the portal all saved in the database no calculate needed and the API to list artifacts support pagination.

**Comment by steven-zou** [comment]
@jcambronero 

We built a patched chartmuseum image `goharbor/chartmuseum-photon:v0.9.0-1.8.0-SP1` based on V1.8 and latest chartmuseum code. We have tested there will be some improvements in the related chart listing API. If it's possible and you're willing to, you can have a try with this patched image to see if the issue can be eliminated. 

Of course, we're planning to deliver a fix of this issue in the next patch release. Pls stay tuned.

**Comment by jcambronero** [comment]
Thank you @steven-zou. We're testing this today.
I haven't tested the upgrade yet, but when loading the project with the large amount of charts from the Harbor UI, I'm still seeing these error messages in the core logs:

```
2020-07-09T17:32:38Z [ERROR] [/chartserver/client.go:124]: GET '/<project_name>/index.yaml' failed with error: Get http://harbor-upgrade-harbor-chartmuseum/<project_name>/index.yaml: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-07-09T17:32:38Z [ERROR] [/core/api/project.go:440]: Failed to get total of charts under project <project_name>: Get http://harbor-upgrade-harbor-chartmuseum/<project_name>/index.yaml: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
2020-07-09T17:32:38Z [ERROR] [/common/api/base.go:229]:
2020-07-09T17:32:38Z [ERROR] [/common/api/base.go:69]: GET /api/projects?page=1&page_size=15 failed with error: {"code":500,"message":"Internal Server Error"}
```

**Comment by jcambronero** [comment]
Hi @steven-zou, no luck with the upgrade unfortunately. We're seeing the same problem.
I'm wondering, will it be possible to get a patched 1.9 version of the harbor-core image with an increased HTTP client timeout setting, 30 seconds or more?

**Comment by steven-zou** [comment]
OK, got it. Anyway, thanks a lot for the verification. As the workaround does not work, I think we need to introduce a fix in the upcoming patch release 2.0.2 that may be released at end of this month(July) from the two perspectives:

 - make sure the chartmuseum binary contain the improved latest diff function
 - adjust the client timeout properties to support calling a low-perf HTTP API

**Comment by steven-zou** [comment]
> I'm wondering, will it be possible to get a patched 1.9 version of the harbor-core image with an increased HTTP client timeout setting, 30 seconds or more?

@jcambronero We can build a dev image for you to do the verification later.

**Comment by reasonerjt** [comment]
This may be a hard limit due to the implementation of ChartMuseum even after the PRs are merged.

We'll test the perf if the charts are stored in OCI registry

**Comment by jcambronero** [comment]
> > I'm wondering, will it be possible to get a patched 1.9 version of the harbor-core image with an increased HTTP client timeout setting, 30 seconds or more?
> 
> @jcambronero We can build a dev image for you to do the verification later.

Thank you @steven-zou!

**Comment by ninjadq** [comment]
@jcambronero 
we've uploaded patched image on docker hub
`docker push goharbor/harbor-core:1.9.0-SP1` for 1.9.0's core with timeout `30` seconds
`goharbor/chartmuseum-photon:v0.9.0-1.9.0-SP1` for chartmuseum with performance improve

**Comment by rafal-prasal** [comment]
just recompiled harbor 1.10.3 tag with the changes
seems that api requests to not fail so much.

--- a/src/chartserver/client.go
+++ b/src/chartserver/client.go
@@ -14,9 +14,9 @@ import (
 )
 
 const (
-       clientTimeout         = 10 * time.Second
-       maxIdleConnections    = 10
-       idleConnectionTimeout = 30 * time.Second
+       clientTimeout         = 60 * time.Second
+       maxIdleConnections    = 60
+       idleConnectionTimeout = 60 * time.Second
 )
 
 // ChartClient is a http client to get the content from the external http server
@@ -36,6 +36,7 @@ func NewChartClient(credential *Credential) *ChartClient { // Create http client
                Transport: &http.Transport{
                        MaxIdleConns:    maxIdleConnections,
                        IdleConnTimeout: idleConnectionTimeout,
+                       ResponseHeaderTimeout: clientTimeout,
                },
        }

**Comment by jcambronero** [comment]
> @jcambronero
> we've uploaded patched image on docker hub
> `docker push goharbor/harbor-core:1.9.0-SP1` for 1.9.0's core with timeout `30` seconds
> `goharbor/chartmuseum-photon:v0.9.0-1.9.0-SP1` for chartmuseum with performance improve

Thank you @ninjadq. We tested this today and noticed the following:
- With both the chartmuseum and harbor-core patched images the upgrade process was completed but it was done really fast (in a matter of seconds) and somehow all the charts were missing after the upgrade.
- With only the harbor-core patched image the upgrade process took much longer (about 15 minutes) and it was successfully completed. We could see all the charts after the upgrade just fine.

We are planning on doing the production upgrade with only the harbor-core patched version (and skip the chartmuseum patch for now). 
Are you planning on releasing new production-ready images with the timeout increase fix for harbor v1.9 and v1.10?

**Comment by steven-zou** [comment]
> > @jcambronero
> > we've uploaded patched image on docker hub
> > `docker push goharbor/harbor-core:1.9.0-SP1` for 1.9.0's core with timeout `30` seconds
> > `goharbor/chartmuseum-photon:v0.9.0-1.9.0-SP1` for chartmuseum with performance improve
> 
> Thank you @ninjadq. We tested this today and noticed the following:
> 
> * With both the chartmuseum and harbor-core patched images the upgrade process was completed but it was done really fast (in a matter of seconds) and somehow all the charts were missing after the upgrade.
> * With only the harbor-core patched image the upgrade process took much longer (about 15 minutes) and it was successfully completed. We could see all the charts after the upgrade just fine.
> 
> We are planning on doing the production upgrade with only the harbor-core patched version (and skip the chartmuseum patch for now).
> Are you planning on releasing new production-ready images with the timeout increase fix for harbor v1.9 and v1.10?

I think at least we'll consider introducing a fix in 1.10.

**Comment by jcambronero** [comment]
Ok, thank you @steven-zou and @ninjadq for all your help!

**Comment by reasonerjt** [comment]
Considering the fix in chartmuseum is not likely to be merged quickly in v2.0.2 we'll only prolong the timeout of the client in Harbor.

There will still be issues if too many charts are stored in one project.

**Comment by ninjadq** [comment]
PR Prolong the timeout in 2.0.2 is https://github.com/goharbor/harbor/pull/12587
PR Prolong the timeout in 1.10.5 is https://github.com/goharbor/harbor/pull/12588

**Comment by jcambronero** [comment]
hi @ninjadq, just wondering if you have an ETA for the v1.10 and v2.0 releases with these fixes (increased core timeout and chartmuseum performance improve)? Thanks!

**Comment by jcambronero** [comment]
hi @ninjadq , recently we have been facing a couple of new disruptive issues. The overall recommendation for both of these is to upgrade to v2.0+. We would like to proceed with the upgrade ASAP but not having the increased timeout for loading the charts is a showstopper for us. Could you please help us expediting a patched 2.0 core version?
The 2 new bugs that we're facing are: [#10296](https://github.com/goharbor/harbor/issues/10296) and [#6515](https://github.com/goharbor/harbor/issues/6515) 

Thanks for your help
